File: workflows/internal/workflow/user_input.go
package workflow

import (
	"time"

	"go.temporal.io/sdk/log"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"

	"agent-harness/workflows/internal/types"
)

// UserInputResponseSignalName is the signal a UserInputRequestWorkflow waits
// on. Distinct from NewMessageSignalName — a response to a specific pending
// request is never folded into a turn's next reasoning step the way an
// ordinary chat message is; it's routed by SignalWorkflow directly against
// this workflow's own execution, correlated by UserInputRequest.RequestID
// (docs/components/user-input.md, "Resolved: Core Mechanism").
const UserInputResponseSignalName = "UserInputResponse"

// UserInputRequestTimeout — docs/components/user-input.md. A pending
// request waits at most this long before being treated as expired. For
// permission gating specifically this is fail-closed: an unanswered
// approval is treated as denied, never as approved by default. A concrete
// value given directly, not guessed — contrast every other numeric
// threshold in this project (compression thresholds, heartbeat intervals),
// deliberately left unsized pending real usage data.
const UserInputRequestTimeout = 1 * time.Hour

// UserInputRequestWorkflow is the reusable primitive docs/components/user-input.md
// resolves: durably wait for a human's response to a request, however long
// that takes (up to UserInputRequestTimeout), and resume exactly there — a
// workflow-level signal wait, not an activity-level block, since an activity
// can't durably block for potentially an hour (see that doc's "Resolved: Why
// an Activity Can't Do This, a Workflow Can"). Kind-agnostic in general;
// ApprovalGatedCall on the input is permission gating's own opt-in use of
// it — an approval request IS a user input request, not a separate workflow
// type layered on top.
//
// Three ways this wait ends, all handled explicitly:
//  1. The response signal arrives — the intended path.
//  2. UserInputRequestTimeout elapses — expired, fail-closed for permission
//     gating (treated as denied).
//  3. This workflow is cancelled (e.g. the parent turn was interrupted by an
//     unrelated new message while this call was still pending — a real,
//     deliberate resolution of a previously-open question: an unrelated
//     message DOES cancel a pending request here, same as it already
//     cancels any other in-flight tool call). Found and fixed alongside this
//     rewrite: an earlier version of this workflow logged this case but
//     never actually updated the pending Postgres row — left it stuck at
//     'pending' forever. Fixed by routing all three exits through the same
//     CloseUserInput activity call.
func UserInputRequestWorkflow(ctx workflow.Context, input types.UserInputRequestWorkflowInput) (types.UserInputRequestWorkflowOutput, error) {
	logger := workflow.GetLogger(ctx)
	workflowID := workflow.GetInfo(ctx).WorkflowExecution.ID
	req := input.Request

	ao := workflow.ActivityOptions{StartToCloseTimeout: activityTimeoutTierA}
	actx := workflow.WithActivityOptions(ctx, ao)
	if err := workflow.ExecuteActivity(actx, "RequestUserInput", req, workflowID).Get(actx, nil); err != nil {
		return types.UserInputRequestWorkflowOutput{}, err
	}
	dispatchInterimDelivery(ctx, input, logger)

	signalChan := workflow.GetSignalChannel(ctx, UserInputResponseSignalName)
	timerCtx, cancelTimer := workflow.WithCancel(ctx)
	timer := workflow.NewTimer(timerCtx, UserInputRequestTimeout)

	var response types.UserInputResponse
	expired := false

	sel := workflow.NewSelector(ctx)
	sel.AddReceive(signalChan, func(c workflow.ReceiveChannel, more bool) {
		c.Receive(ctx, &response)
	})
	sel.AddFuture(timer, func(f workflow.Future) {
		expired = true
		logger.Info("UserInputRequestWorkflow expired waiting for a response", "request_id", req.RequestID)
	})
	sel.AddReceive(ctx.Done(), func(c workflow.ReceiveChannel, more bool) {
		logger.Info("UserInputRequestWorkflow cancelled while waiting", "request_id", req.RequestID)
	})
	sel.Select(ctx)
	cancelTimer()

	cancelled := ctx.Err() != nil

	// Best-effort Postgres close-out — same tolerance as every other
	// end-of-turn bookkeeping call in this codebase (Persist/Deliver are
	// also fire-and-forget on their own errors). Uses a fresh, non-cancelled
	// activity context deliberately: if this workflow itself was cancelled,
	// ctx is already done, and an activity dispatched under it would never
	// even get scheduled.
	closeStatus := "answered"
	if expired || cancelled {
		closeStatus = "cancelled"
	}
	bg, cancelBg := workflow.NewDisconnectedContext(ctx)
	defer cancelBg()
	cao := workflow.ActivityOptions{StartToCloseTimeout: activityTimeoutTierA}
	cactx := workflow.WithActivityOptions(bg, cao)
	_ = workflow.ExecuteActivity(cactx, "CloseUserInput", req.RequestID, closeStatus, response.SelectedOptionID, response.FreeText).Get(cactx, nil)

	if cancelled {
		// Found via live testing: an earlier version returned here directly,
		// which correctly closed out user_input_requests (the call above)
		// but skipped markDenied entirely — leaving tool_calls stuck at
		// 'pending' forever for this specific path, the same class of bug
		// just moved one table over. Uses bg (disconnected from the
		// already-cancelled ctx) for the same reason CloseUserInput does.
		if input.ApprovalGatedCall != nil {
			_ = markDenied(bg, input.ApprovalGatedCall.ToolCallID, "cancelled")
		}
		return types.UserInputRequestWorkflowOutput{}, ctx.Err()
	}

	out := types.UserInputRequestWorkflowOutput{Response: response}

	if input.ApprovalGatedCall == nil {
		return out, nil
	}

	approved := !expired && response.SelectedOptionID != nil && *response.SelectedOptionID == "approve"
	if !approved {
		reason := "denied_by_user"
		if expired {
			reason = "expired"
		}
		_ = markDenied(bg, input.ApprovalGatedCall.ToolCallID, reason)
		toolOut := types.ToolCallOutput{ToolCallID: input.ApprovalGatedCall.ToolCallID, Status: "cancelled"}
		out.ToolCallOutput = &toolOut
		return out, nil
	}

	// Approved — dispatch the real ToolCall activity, same timing/retry
	// policy any other call to this tool_name would get (tool_tiers.go).
	timing := toolTimingFor(input.ApprovalGatedCall.ToolName)
	tao := workflow.ActivityOptions{
		ActivityID:          input.ApprovalGatedCall.ToolCallID,
		StartToCloseTimeout: timing.StartToCloseTimeout,
		HeartbeatTimeout:    timing.HeartbeatTimeout,
		WaitForCancellation: true,
		RetryPolicy:         &temporal.RetryPolicy{MaximumAttempts: 3},
	}
	tactx := workflow.WithActivityOptions(ctx, tao)
	var toolOut types.ToolCallOutput
	if err := workflow.ExecuteActivity(tactx, "ToolCall", types.ToolCallInput{ToolCallID: input.ApprovalGatedCall.ToolCallID}).Get(tactx, &toolOut); err != nil {
		toolOut = types.ToolCallOutput{ToolCallID: input.ApprovalGatedCall.ToolCallID, Status: statusFromError(err)}
	}
	out.ToolCallOutput = &toolOut
	return out, nil
}

// connectionInterimDeliveryActivity mirrors connectionDeliveryActivity/
// connectionDeliveryChunkActivity (turn.go) — same literal-lookup-not-a-
// registry reasoning (two platforms exist today, a third is a one-line
// addition). A separate lookup, not folded into either existing one:
// pushing a pending request's prompt is a different activity, with a
// different signature (takes request_id, not turn_id — the interim-
// delivery activity reads prompt/options/channel routing off
// user_input_requests itself, since that row already exists by the time
// this is ever called), from either a turn's final answer (Deliver) or a
// streamed chunk of one (DeliverChunk).
func connectionInterimDeliveryActivity(platform string) (activityName string, timeout time.Duration, ok bool) {
	switch platform {
	case "discord":
		return "DiscordDeliverInterim", activityTimeoutTierA, true
	case "discord-voice":
		return "VoiceDeliverInterim", 5 * time.Minute, true
	default:
		return "", 0, false
	}
}

// dispatchInterimDelivery — docs/components/user-input.md's "Mid-turn
// interim delivery" (push half, A+B). Best-effort, fire-and-forget from
// this workflow's own perspective: a failure to PUSH the prompt must never
// fail the whole UserInputRequestWorkflow (the request is already durably
// recorded in Postgres by RequestUserInput above, and Web's own poll-based
// surfacing — gateway/web.md's handlePoll — doesn't depend on this push at
// all) — logged and swallowed, same tolerance this codebase already gives
// Persist/Deliver's own best-effort bookkeeping calls.
//
// No connectionID means either Web (no connection-lease concept, already
// solved by polling) or a plain decision-request consumer that never set
// these fields — either way, nothing to push, silently a no-op rather than
// an error, matching turn.go's own deliverConnectionBased short-circuit for
// the same empty-connectionID case.
//
// Deliberately NOT raced against the response wait the way
// deliverConnectionBased races final delivery against an interrupt: an
// interim push is a fire-and-forget side effect with its own idempotency
// (prompt_delivered_at), not something that needs to be cancelled if a
// response arrives first — by the time it would matter, the send has
// either already gone out or the activity call itself is short-lived
// (activityTimeoutTierA/near-instant for text, bounded for voice TTS).
func dispatchInterimDelivery(ctx workflow.Context, input types.UserInputRequestWorkflowInput, logger log.Logger) {
	if input.ConnectionID == "" {
		return
	}
	platform := platformFromSessionKey(input.SessionKey)
	activityName, timeout, ok := connectionInterimDeliveryActivity(platform)
	if !ok {
		return
	}
	ao := workflow.ActivityOptions{
		StartToCloseTimeout: timeout,
		TaskQueue:           "deliver:" + platform + ":" + input.ConnectionID,
	}
	actx := workflow.WithActivityOptions(ctx, ao)
	if err := workflow.ExecuteActivity(actx, activityName, input.Request.RequestID).Get(actx, nil); err != nil {
		logger.Warn("dispatchInterimDelivery: failed to push pending request prompt", "request_id", input.Request.RequestID, "error", err)
	}
}

// markDenied is a small, best-effort helper — the tool_calls row was minted
// by ModelCall with status='pending' (see tool_calls schema note on why that
// default exists); if this call never reaches the real ToolCall activity at
// all (denied, expired, or cancelled before approval), nothing else will
// ever transition it out of 'pending' otherwise.
func markDenied(ctx workflow.Context, toolCallID string, reason string) error {
	ao := workflow.ActivityOptions{StartToCloseTimeout: activityTimeoutTierA}
	actx := workflow.WithActivityOptions(ctx, ao)
	return workflow.ExecuteActivity(actx, "DenyToolCall", toolCallID, reason).Get(actx, nil)
}

File: activities/activities/user_input.py
"""UserInputRequestWorkflow's two activities — docs/components/user-input.md.
Everything content-bearing about a pending request lives in Postgres, read
back by the workflow only as the {status, response} it needs to decide
whether it's still waiting; the request/response payloads themselves never
round-trip through the workflow beyond what UserInputRequestWorkflow already
holds as its own input/output (reference-passing contract,
components/temporal-workflow.md).

Two request kinds share this primitive: `permission` (approval gating — the
tool_calls row is transitioned by ToolCall/DenyToolCall) and `question`
(`ask_user`, docs/components/turn-pipeline.md — turn.go dispatches this as a
child workflow and parks the loop on it, and the answer is written back into
the ask_user tool_calls row here so the next ModelCall sees it as an
observation).
"""

from __future__ import annotations

import json
import logging

from temporalio import activity

from .types import UserInputRequest

logger = logging.getLogger(__name__)


class RequestUserInputActivity:
    def __init__(self, pool):
        self._pool = pool

    @activity.defn(name="RequestUserInput")
    async def __call__(self, request: UserInputRequest, workflow_id: str) -> None:
        prompt = request.prompt
        options = [{"id": o.id, "label": o.label} for o in request.options]

        # ask_user (Kind "question"): turn.go passes only the request id (== the
        # ask_user tool_call_id) — the question + options live in the model's
        # own tool_calls.arguments, read here so the workflow never holds them.
        # Safe option ids (opt_N) since a model-authored option string can
        # contain ':' and would break discord_user_input.go's custom_id parse.
        if request.kind == "question" and not prompt:
            tc = await self._pool.fetchrow(
                "SELECT arguments FROM tool_calls WHERE tool_call_id = $1", request.request_id
            )
            args = json.loads(tc["arguments"]) if tc and tc["arguments"] else {}
            prompt = str(args.get("question", "")).strip() or "(the assistant needs your input)"
            options = [
                {"id": f"opt_{i}", "label": str(o).strip()}
                for i, o in enumerate(args.get("options") or [])
                if str(o).strip()
            ]

        await self._pool.execute(
            "INSERT INTO user_input_requests "
            "(request_id, turn_id, workflow_id, kind, prompt, options, allow_free_text, context) "
            "VALUES ($1, $2, $3, $4, $5, $6, $7, $8) "
            "ON CONFLICT (request_id) DO NOTHING",
            request.request_id,
            request.turn_id,
            workflow_id,
            request.kind,
            prompt,
            json.dumps(options),
            request.allow_free_text,
            json.dumps(request.context),
        )
        logger.info(
            "RequestUserInput[%s]: kind=%s %r options=%r",
            request.request_id, request.kind, prompt[:80], [o["label"] for o in options],
        )


class CloseUserInputActivity:
    """Closes out a pending request in exactly one of two terminal states —
    'answered' (a real response arrived) or 'cancelled' (expired, or the
    workflow was cancelled by an unrelated interrupt). UserInputRequestWorkflow
    calls this on all of its own exit paths.

    For an `ask_user` (kind 'question') it ALSO transitions the model's own
    ask_user tool_calls row (request_id == tool_call_id): 'answered' → status
    'ok' with {"answer": ...} so the next ModelCall reads it as an observation;
    otherwise 'cancelled' (the user typically replied with a plain message
    instead, which turn.go folds in as the next user turn)."""

    def __init__(self, pool):
        self._pool = pool

    @activity.defn(name="CloseUserInput")
    async def __call__(
        self, request_id: str, status: str, selected_option_id: str | None, free_text: str | None
    ) -> None:
        row = await self._pool.fetchrow(
            "UPDATE user_input_requests SET status = $2, selected_option_id = $3, "
            "free_text = $4, answered_at = now() WHERE request_id = $1 RETURNING kind, options",
            request_id,
            status,
            selected_option_id,
            free_text,
        )
        logger.info("CloseUserInput[%s]: status=%s selected=%r", request_id, status, selected_option_id)

        if not row or row["kind"] != "question":
            return

        answer = (free_text or "").strip() or None
        if answer is None and selected_option_id is not None:
            opts = json.loads(row["options"]) if row["options"] else []
            answer = next((o["label"] for o in opts if o["id"] == selected_option_id), selected_option_id)

        if status == "answered" and answer:
            await self._pool.execute(
                "UPDATE tool_calls SET status = 'ok', result = $2, side_effect = 'none', "
                "completed_at = now() WHERE tool_call_id = $1 AND status = 'pending'",
                request_id,
                json.dumps({"answer": answer}),
            )
        else:
            await self._pool.execute(
                "UPDATE tool_calls SET status = 'cancelled', reason = 'no_button_answer', "
                "side_effect = 'none', completed_at = now() WHERE tool_call_id = $1 AND status = 'pending'",
                request_id,
            )
